src/ringmesh/geomodel/geomodel_builder.cpp

```diff
@@ -995,8 +1022,336 @@ namespace RINGMesh
+                ++surface_counter;
+            }
+        }
+
+        void check_internal_floating_lines(
+            const std::vector< LineIncidentSurfacePair >&
+                line_indicent_surfaces,
+            const index_t nb_found_surfaces )
+        {
+            std::vector< bool > are_surfaces_hole( nb_found_surfaces, true );
+            for( auto line_id : range( geomodel_.nb_lines() ) )
+            {
+                if( line_indicent_surfaces[line_id].plus_surface_index()
+                    != line_indicent_surfaces[line_id].minus_surface_index() )
+                {
+                    are_surfaces_hole[line_indicent_surfaces[line_id]
+                                          .plus_surface_index()] = false;
+                    are_surfaces_hole[line_indicent_surfaces[line_id]
+                                          .minus_surface_index()] = false;
+                }
+            }
+            index_t nb_pb_surfaces{ static_cast< index_t >( std::count(
+                are_surfaces_hole.begin(), are_surfaces_hole.end(), true ) ) };
+            if( nb_pb_surfaces > 0 )
+            {
+                throw RINGMeshException( "Surface2D",
+                    "During surface from corners "
+                    " and lines build, ",
+                    nb_pb_surfaces,
+                    " group(s) of lines are "
+                    "floating inside a surface. This is not yet handled by the "
+                    "algorithm. Aborting..." );
+            }
+        }
+
+        std::vector< vec2 > get_surface_polygon_vertices(
+            const std::vector< OrientedLine >& surface_boundaries )
+        {
+            std::vector< vec2 > polygon_vertices;
+            for( auto cur_surf_boundary : surface_boundaries )
+            {
+                const auto& cur_line =
+                    this->geomodel_.line( cur_surf_boundary.index );
+                for( auto vertex : range( 1, cur_line.nb_vertices() ) )
+                {
+                    polygon_vertices.push_back( cur_line.vertex(
+                        cur_surf_boundary.side
+                            ? vertex
+                            : ( cur_line.nb_vertices() - 1 ) - vertex ) );
+                }
+            }
+            return polygon_vertices;
+        }
+
+        void build_surface_polygons(
+            const std::vector< std::vector< OrientedLine > >&
+                surface_boundary_lines )
+        {
+            for( const auto& surface_boundaries : surface_boundary_lines )
+            {
+                std::vector< vec2 > polygon_vertices =
+                    get_surface_polygon_vertices( surface_boundaries );
+                std::vector< index_t > polygon_corners(
+                    polygon_vertices.size() );
+                std::iota( polygon_corners.begin(), polygon_corners.end(), 0 );
+                GeoModelBuilder2D builder( geomodel_ );
```

Comment:
The builder could be outside the lop or even in the data member of the class.

Reply:
Or better, take it by reference from the GMBuilder call (for example impl_->foo(*this, parameters...))